### PR TITLE
ssd1306 more functions

### PR DESCRIPTION
**Pull request by Zaltora**
Hi, It is a lot of new functions for ssd1306 driver. Adapted from other libs. this PR add:

-An new example.
-2D graphical functions.
-Font system (need more option to choice what to compile, just initial )

Make a new font with this tool: http://www.eran.io/the-dot-factory-an-lcd-font-and-image-generator/
Source: https://github.com/baoshi/ESP-I2C-OLED and some functions from adafruits lib.

### CODE REVIEW

File: extras/fonts/fonts.c [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/*
+ * oled_fonts.c
+ *
+ *  Created on: 8 dec. 2016
+ *      Author: zaltora
+ */
+#include "fonts.h"
+
+extern const font_info_t  glcd_5x7_font_info;
+extern const font_info_t  tahoma_8pt_font_info;
```
Comment (UncleRus) [inline extras/fonts/fonts.c]: We cannot use here in esp-open-rtos proprietary fonts like tahoma.

**Comment by Zaltora** [comment]
It is true, totaly forget about font license. I will search new font to replace them.  Any idea ?
It exist a lot of font. Add most popular open font can be great.

**Comment by UncleRus** [comment]
For example Roboto, Droid Sans. Both distributed under Apache license.

**Comment by Zaltora** [comment]
I used Roboto, Hesited with Lato :) 
Need improve font part and add more. Can be used by other screen library, i guess.

**Comment by UncleRus** [comment]
Now we have simple tool to convert .pcf/.bdf files to source/
Preparing:
```
$ pip3 install pillow jinja2
```
Converting PCF/BDF font to PIL:
```
$ pilfont ter-u16b_iso-8859-1.pcf
```
Generating source:
```
$ ./create_font.py -f ter-u16b_iso-8859-1.pil -n "Terminus" -c iso-8859-1 > ../data/font_terminus_16x32_iso-8859-1.h
```

Proportional fonts are not working, both of Roboto. I didn't try to generate my own proportional font and I believe error can be somewhere in font data, not in the code.

**Comment by Zaltora** [comment]
It is totally what i imagine. a tool, include system, improvements.
Very nice @UncleRus !
I will test it next week.

**Comment by UncleRus** [comment]
And one more TODO: we also need correct and fast clipping in `ssd1306_draw_char()`.

**Comment by Zaltora** [comment]
Just added HW scrolling.
Question about setting system : Where app user define font he want and font he dont want ? (Makefile app file? ,  this own config.h file ? He dont need change config.h or componen.mk from  font directory ?
What is best way to set font or protocol avaible ?

Nice addition @UncleRus  :) .

**Comment by UncleRus** [comment]
We have scrolling now! That's great :)

Any font can be selected or de-selected in the app makefile like this:
```
PROGRAM=SSD1306_fps
EXTRA_COMPONENTS = extras/ssd1306 extras/i2c extras/fonts

FONTS_BITOCRA_6X11 = 1

include ../../common.mk
```
We had to disable all fonts by default in `fonts/defaults.mk` before merging this PR.

**Comment by Zaltora** [comment]
Great, I will review exemple part tomorow ( clear high part screen )  before change font. 
I see that fps drop when font is huge. Maybe need break loop if drawpixel is out of screen

**Comment by Zaltora** [comment]
Now draw a character can return negative value if write outside of the screen.

**Comment by UncleRus** [comment]
@Zaltora Awesome work!
I'm going to squash & merge it after you add licensing information to headers.

**Comment by Zaltora** [comment]
Well, i am new with license. It is good ?
I see just one header to change. ssd1306 part is already ok.

**Comment by UncleRus** [comment]
Excellent!

**Comment by UncleRus** [comment]
Could you please squash this PR?

**Comment by Zaltora** [comment]
I squash all commit from "ssd1306 2D and font" to "Licence fix".
When i  doing this, it is added these commits : 
-Fixes Bug(s)
-Second bug. fix attempt.
-Fix status with bus.
-Readable option

from #302 , i dont no why :( .  i Need force rebase ?  What happen ?

**Comment by Zaltora** [comment]
Thank you, need more training with git :)

**Pull request merged**